### ISSUE TITLE
client.loadMySky  generates ERRORS in  version v4.0.26-beta and v4.0.27-beta ! Help!

### ISSUE THREAD

**Issue by parajbs** [SkynetLabs/skynet-js#428]
Hello skynet-js support team,

I rebuilt skynet-js version v4.0.25-beta / v4.0.26-beta and v4.0.27-beta with webpack! The create ran with all versions
without error!
Then I uploaded all versions to siasky.net!
After that I created a test app on jsfiddle for all 3 versions and also the version of "https://skynet-js.hns.siasky.net/4.0-beta/index.js"!

- The test app in the version of "https://skynet-js.hns.siasky.net/4.0-beta/index.js" is all OK!
- The test app in version v4.0.25-beta is also still OK!
- But in the test apps in version v4.0.26-beta and v4.0.27-beta there are Errors with CORS at COMMAND
 const mySky = await client.loadMySky(hostApp, { debug: true, dev: false, });

Hope someone can help or fix bugs in the new version!

**###########**

Bugs in version v4.0.26-beta and v4.0.27-beta :
At the command:
const mySky = await client.loadMySky(hostApp, { debug: true, dev: false, });

**#########**

Console.log in the browser shows these errors:
**Error 1:**
Access to XMLHttpRequest at 'https://jsfiddle.net/' (redirected from 'https://fiddle.jshell.net/') from origin 'https://fiddle.jshell.net' has been blocked by CORS policy: The value of the 'Access-Control-Allow-Credentials' header in the response is '' which must be 'true' when the request's credentials mode is 'include'. The credentials mode of requests initiated by the XMLHttpRequest is controlled by the withCredentials attribute.

**Error2:**
HEAD https://jsfiddle.net/net::ERR_FAILED 200

**Error3:**
Uncaught (in promise) ExecuteRequestError: Error response did not contain expected field 'response'.
    at Function.From (index.js:2:179510)
    at index.js:2:231842
    at async hi.executeRequest (index.js:2:234421)
    at async hi.resolvePortalUrl (index.js:2:235268)
    at async Function.New (index.js:2:216546)
    at async hi.Oo [as loadMySky] (index.js:2:215231)
    at async init (?editor_console=true:158:19)

**Error4:**
XHR failed loading: HEAD "https://jsfiddle.net/".

**#################**

Here are the links to test:
test 1: ALL OK!
skynet-js v4.0-beta from skynet-js.hns
link: [](https://jsfiddle.net/vqw9rjea/1/)

test 2: ALL OK!
skynet-js v4.0.25-beta
link: [](https://jsfiddle.net/vqw9rjea/2/)

test 3: ERROR! OR BUG!
skynet-js v4.0.26-beta
link: [](https://jsfiddle.net/vqw9rjea/3/)

test 4: ERROR! OR BUG!
skynet-js v4.0.27-beta
link: [](https://jsfiddle.net/vqw9rjea/4/)

Hope someone can help or fix bugs in the new version!
parajbs

**Comment by parajbs** [comment]
Here are the links to test:
test 1: ALL OK!
skynet-js v4.0-beta from skynet-js.hns
link: https://jsfiddle.net/vqw9rjea/1/

test 2: ALL OK!
skynet-js v4.0.25-beta
link: https://jsfiddle.net/vqw9rjea/2/

test 3: ERROR! OR BUG!
skynet-js v4.0.26-beta
link: https://jsfiddle.net/vqw9rjea/3/

test 4: ERROR! OR BUG!
skynet-js v4.0.27-beta
link: https://jsfiddle.net/vqw9rjea/4/

**Comment by m-cat** [comment]
Thanks for the detailed report. I'll take a look!

**Comment by parajbs** [comment]
Here are a few links to files:

v4.0.25 beta:
babel.config.json:
https://fg1144gdnllpin27thh5v2igskv22t8q3832aa9bmj2lo8pvdf20j3g.siasky.net/skynet-js/v4.0.25-beta/babel.config.json
webpack.config.js:
https://fg1144gdnllpin27thh5v2igskv22t8q3832aa9bmj2lo8pvdf20j3g.siasky.net/skynet-js/v4.0.25-beta/webpack.config.js
package.json:
https://fg1144gdnllpin27thh5v2igskv22t8q3832aa9bmj2lo8pvdf20j3g.siasky.net/skynet-js/v4.0.25-beta/package.json

v4.0.26 beta:
babel.config.json:
https://fg1144gdnllpin27thh5v2igskv22t8q3832aa9bmj2lo8pvdf20j3g.siasky.net/skynet-js/v4.0.27-beta/babel.config.json
webpack.config.js:
https://fg1144gdnllpin27thh5v2igskv22t8q3832aa9bmj2lo8pvdf20j3g.siasky.net/skynet-js/v4.0.26-beta/webpack.config.js
package.json:
https://fg1144gdnllpin27thh5v2igskv22t8q3832aa9bmj2lo8pvdf20j3g.siasky.net/skynet-js/v4.0.26-beta/package.json

v4.0.27 beta:
babel.config.json:
https://fg1144gdnllpin27thh5v2igskv22t8q3832aa9bmj2lo8pvdf20j3g.siasky.net/skynet-js/v4.0.27-beta/babel.config.json
webpack.config.js:
https://fg1144gdnllpin27thh5v2igskv22t8q3832aa9bmj2lo8pvdf20j3g.siasky.net/skynet-js/v4.0.27-beta/webpack.config.js
package.json:
https://fg1144gdnllpin27thh5v2igskv22t8q3832aa9bmj2lo8pvdf20j3g.siasky.net/skynet-js/v4.0.26-beta/package.json

Thanks in advance for watching!

**Comment by parajbs** [comment]
hello m-cat,

I don't have an hns domain at the moment.
But the app would not only operate via .hns.siasky.net but also its own .hns.owndomain.com own tld domain or .hns.to (decentralized)!

It would therefore also be important and more decentralized if the app were also stored / worked outside of siasky.net and still worked with client.loadMySky there as a decentralized login mysky / skydb and skyregistry!

Would also help developers who e.g. Test and develop your app on jsfiddle or codepen!

it would also be cool if you check the files "babel.config.json" and "webpack.config.js" and "package.json (as package-webpack.json)" and include them in your skynet-js repository and maybe one too Create readme-webpack-skynet-js-4.x!
Then other v4.0.x beta developers could also work with the current skynet-js-beta version and not just with the version of "https://skynet-js.hns.siasky.net/4.0-beta/index .js" (I'm not sure how up-to-date it is, that's why it's the way it is.)!

#####

I have another question:
I myself am looking for an approach, whether and how you can register with Metamask(web3.ethereum), at skynet-mysky and then log in at skynet-mysky!
Then you could log in to Metamask, and would automatically log in to skynet-mysky!
But I think skynet-mysky-login must be expanded so that it works!

Maybe there are already approaches, would definitely increase the group of developers enormously and also the group of customers for skapps (login will be even easier for customers)!

Let's see what I can find for approaches.

**Comment by m-cat** [comment]
> It would therefore also be important and more decentralized if the app were also stored / worked outside of siasky.net and still worked with client.loadMySky there as a decentralized login mysky / skydb and skyregistry!

Yes, if the app is not on a portal then there's no way we can redirect to a different portal. So we should not try to do that, which should fix your issue. I'm going to get to this ASAP.

> it would also be cool if you check the files "babel.config.json" and "webpack.config.js" and "package.json (as package-webpack.json)" and include them in your skynet-js repository and maybe one too Create readme-webpack-skynet-js-4.x!
Then other v4.0.x beta developers could also work with the current skynet-js-beta version and not just with the version of "https://skynet-js.hns.siasky.net/4.0-beta/index .js" (I'm not sure how up-to-date it is, that's why it's the way it is.)!

I don't understand this part.

> I have another question:
I myself am looking for an approach, whether and how you can register with Metamask(web3.ethereum), at skynet-mysky and then log in at skynet-mysky!
Then you could log in to Metamask, and would automatically log in to skynet-mysky!
But I think skynet-mysky-login must be expanded so that it works!

Yes, we intended for MySky login to be integrated with third party services and built in some support for that early on. However, this responsibility is going to move into the [kernel](https://github.com/SkynetLabs/skynet-kernel) soon.

**Comment by m-cat** [comment]
Hey @parajbs, I've raised a PR that should fix this.

https://github.com/SkynetLabs/skynet-js/pull/431

Can you please let me know what command you used to build a standalone script in those fiddles?

**Comment by parajbs** [comment]
Hey m-cat,

Here is my description and the files!
Again it would be cool if you include the description and files in repo of skynet-js! You can rewrite the readme to suit you!


Instructions to the build skynet-js with webpack!

1. git clone --branch v4.0.x-beta https://github.com/SkynetLabs/skynet-js.git
   ATTENTION! The x in v4.0.x-beta must be replaced with the version number (e.g.: 25 or 27)!

2. Download the file Build_skynet-js_webpack.zip from:
   https://000cgm1d3jqlflm1msidupo6pluu4ia5dllh8aipt5hh89ljk2jij6g.siasky.net/zip/Build_skynet-js_webpack.zip

3. Unzip the zip file and copy the 4 files (babel.config.json / webpack.config.js / package-webpack-v4_0_25-beta.json / and README_build_skynet-js_with_webpack .txt) goes directly to the "skynet-js" directory

4. cd skynet-js

5. Rename the file "package.json" to "package-org.json"! 
   Then rename the file "package-webpack-v4_0_25-beta.json" to "package.json".

6. Here we go! Here are the other commands:

method1:
6.1_a yarn

6.1_b yarn build

6.1_c yarn run webpack


method2:
6.2_a npm install

6.2_b npm build

6.2_c npx webpack


Description of package-webpack-v4_0_25-beta.json
It was added under "dependencies"
 "@babel/runtime": "^7.11.2",

And added under "devDependencies"
    "@babel/cli": "^7.11.6",
    "@babel/core": "^7.11.6",
    "@babel/plugin-proposal-class-properties": "^7.10.4",
    "@babel/plugin-transform-runtime": "^7.11.5",
    "@babel/preset-env": "^7.11.5",
    "@babel/preset-typescript": "^7.10.4",
    "babel-loader": "^8.2.2",
    "babel-plugin-transform-class-properties": "^6.24.1",
    "web pack": "^5.23.0",
    "webpack-cli": "^4.5.0",
    "webpack-merge": "^5.7.3",


I hope I've described everything correctly, let me know if you have any problems! I myself build it on Windows10!

**Comment by parajbs** [comment]
hello m-cat,

I created the branch "" from skynet-js with webpack, yes
unfortunately still the same problem with client.loadMySky CORS!

Here is the link to test5:
https://jsfiddle.net/Lwhkv86e/

and here the new skynet-js:
https://fg35i8472fvv3bjafqgfrqer5pecsmmoppiee9n9qn3j1aj8pafg7ng.siasky.net/skynet-js/v4.0.27-beta-no-portal-no-redirect/index.js

Hopefully the problem can be found and eliminated because, as I said, everything was ok with version v4.0.25-beta!

I'm also looking for this bug in the code myself!
I'll take a look at the changes after version v4.0.25-beta to v4.0.26-beta, where the CORS bug with client.loadMySky appears for the first time!

If you have any ideas where I can look, let me know and I'll test it myself!

**Comment by m-cat** [comment]
What is the error you're getting? I'm seeing `"MySky loading end"` this time.

**Comment by m-cat** [comment]
@parajbs I see the error in the console. That's probably because `skynet-js` tries to make a request to `jsfiddle` to check if it's a portal. If any error occurs, it is caught and `skynet-js` proceeds under the assumption that it's not a portal. So this all seems to be working as expected now.

**Comment by parajbs** [comment]
hello m-cut,

sorry, i saw that the message "MySky loading end" appears in the browser log after the CORS bug!
But I had only read 2 errors in red letters and thought it was the same problem, but apparently there is a different problem!

I think both errors ("Access to XMLHttpRequest " and "net::ERR_FAILED 200") there!
But after researching and testing for a long time, I got the error at jsfiddle.net and also at codepen.io!
When I tested it at jsbin.com, I was surprised!
Because there were no errors at jsbin.com and everything went OK!

Now to the errors:
at jsfiddle.net:
The app is loaded in the browser via the url e.g.: https://jsfiddle.net/Lwhkv86e/ , but the white jsfiddle output field (bottom right) appears to be an iframe or something similar, since it uses the url "fiddle.jshell. net" and not the url "jsfiddle.net"!
Or the other way around with the urls, I'm not quite sure!
I think there are two different main domains in the header!
Therefore the errors!

at codepen.io:
The app is loaded in the browser via the url e.g.: https://codepen.io/pen/, but the white output field (on the right) appears to be an iframe or similar, since it is accessed via the url "https://cdpn. io/" and not the url "https://codepen.io/pen/"!
I'm not entirely sure about this either!
Here, too, I think there are two different main domains in the header!
Therefore the errors as with jsfiddle.net!


at jsbin.com
I saw that the app is loaded via the url "https://jsbin.com", a kind of iframe is also used here, but the output comes via a subdomain url "https://null.jsbin.com/ " or "https://static.jsbin.com" !
So the same main domain is in the header and everything is OK! There are no mistakes!

at my home on the server the app is also OK, because I don't load an iframe but load an index.html via my domain!


my result:
So the errors are there, but only on some online editors where two different main domains are used or on pages that use an iframe with different main domains!
What exactly and where the errors in the code are caused, I haven't found out yet!

**Comment by m-cat** [comment]
Yeah, looks like the jsfiddle iframe that actually runs the code lives in a different domain. When `skynet-js` tries to query the current URL to check if it's a portal, it queries the current URL which is `jsfiddle.com`, but since the request comes from the iframe at `fiddle.jshell.net` there is a CORS error.

I don't think there is anything we can do about that. We don't really expect skapps to be running from code editors like jsfiddle. The main use case is that they are uploaded to Skynet portals.

The main reason I wanted to support non-portal domains is because of the use case of pointing a custom domain to a skapp via dnslink. For example, it's possible to have a skapp at `<skapp>.siasky.net` and point to it from `skapp.to` and access your skapp at `skapp.to`. 

Everything seems to be working, and I'm not worried about the CORS errors which are expected. I will close the issue, but please open a new issue if you come across anything else unexpected. Thank you so much for the report and for the detailed analysis -- it was extremely useful!   🙌
